Gems/Atom/Feature/Common/Code/Source/RayTracing/RayTracingRingBuffer.h

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <Atom/RHI.Reflect/Format.h>
+#include <Atom/RHI.Reflect/FrameCountMaxRingBuffer.h>
+#include <Atom/RPI.Public/Buffer/Buffer.h>
+#include <Atom/RPI.Public/Buffer/BufferSystemInterface.h>
+#include <AzCore/std/containers/span.h>
+#include <AzCore/std/string/string.h>
+
+namespace AZ::Render
+{
+    //! A class which manages a FrameCountMax number of RPI buffers and manages them in a ring buffer structure, meaning that whenever data
+    //! needs to be updated, the current buffer index is incremented (mod FrameCountMax) and the data is then written to the new current
+    //! buffer, such that the other buffers stay valid.
+    class RayTracingRingBuffer : public RHI::FrameCountMaxRingBuffer<Data::Instance<RPI::Buffer>>
```

Comment:
1. RPI does have a ring buffer like implementation. Take a look at RPI::DynamicBuffer and RPI::DynamicBufferAllocator. I wonder if DynamicBuffer and this buffer can leverage each other.
2. Any reason why this is called RayTracingRingBuffer even though it has no RT functionality built in. Why not just call it RingBuffer?

Reply:
1. I could see DynamicBufferAllocator using RingBuffer as a base class, the other way round would not work well since in contrast to the usages of DynamicBuffer (ImGui and AuxGeometry) it would not be acceptable if eg. RT-data is not uploaded completely if the buffer runs out of space in game mode, and resizing a single RHI buffer if there is too much data would invalidate the old data. The differences would be:
   1. Three RPI buffers instead of one
   2. Using UpdateCurrentBufferData (Buffer::UpdateData) instead of Map/Unmap: This would lead to more calls to map/unmap, unless uploading data is handled differently based on the HeapMemoryLevel (keep the buffer mapped if HeapMemoryLevel::Host is used)
4. There was not really a reason for naming it RayTracingRingBuffer other than that is is used only in ray-tracing related buffers, which is also not entirely true since it is used in DiffuseProbeGridFeatureProcessor; but this would also not be a good reason for naming it like this. Would it then make sense to move it to RPI.Public/Buffer, Feature/Utils or leave it in Feature/RayTracing?

Reply:
I renamed the class to `RingBuffer` and moved it to `Atom/RPI.Public/Buffer/ ` such that it can be used by DynamicBufferAllocator.
I changed `DynamicBufferAllocator` to use RingBuffer for allocations, which simplifies the code in DynamicBufferAllocator a lot. I did not use inheritance there but composition, since the interface of DynamicBufferAllocator is completely different and keeping the interface unchanged seemed to make sense.
What I found a bit strange is that the function `RHI::AlignUp` was used to change the current allocation size, but shouldnt it be used for the current allocation start address instead?
What are your thoughts on this @moudgils ?

Reply:
Looks good to me. I am going to quickly ping @VickyAtAZ to do a pass as well.